service/api_task.go

```diff
@@ -568,13 +553,34 @@ func assignNextAvailableTask(ctx context.Context, taskQueue *model.TaskQueue, di
 		})
 
 		if ok && isTaskGroupNewToHost(currentHost, nextTask) {
-			numHosts, err := host.NumHostsByTaskSpec(nextTask.TaskGroup, nextTask.BuildVariant, nextTask.Project, nextTask.Version)
-			if err != nil {
-				return nil, false, errors.WithStack(err)
+			dispatchRace := ""
+			minTaskGroupOrderNum := 0
+			if nextTask.TaskGroupMaxHosts == 1 {
+				// regardless of how many hosts are running tasks, if this host is running the earliest task in the task group we should continue
+				minTaskGroupOrderNum, err = host.MinTaskGroupOrderRunningByTaskSpec(nextTask.TaskGroup, nextTask.BuildVariant, nextTask.Project, nextTask.Version)
+				if err != nil {
+					return nil, false, errors.WithStack(err)
+				}
+				// if minTaskGroupOrderNum is 0 then some host doesn't have order cached, revert to previous logic
+				if minTaskGroupOrderNum != 0 && minTaskGroupOrderNum < nextTask.TaskGroupOrder {
+					dispatchRace = fmt.Sprintf("current task is order %d but another host is running %d", nextTask.TaskGroupOrder, minTaskGroupOrderNum)
+				}
+			}
+			// for multiple-host task groups and single-host task groups without order cached
+			if minTaskGroupOrderNum == 0 {
```

Comment:
Should this read `if minTaskGroupOrderNum == 0 | nextTask.TaskGroupMaxHosts > 1`? Otherwise, multi-host task groups with cached task groups order numbers will not enter this branch.

Reply:
Only single host task groups go through line 558, so minTaskGroupOrderNum == 0 is true for multi-host task groups.